rfcs/20240919-dnncompat/README.md

```diff
@@ -0,0 +1,936 @@
+Status_t MatmulPreferenceCreate(MatmulPreference_t* pref);
+Status_t MatmulPreferenceSetAttribute(MatmulPreference_t pref,
+                                      MatmulPreferenceAttributes_t attr,
+                                      const void *buf,
+                                      size_t sizeInBytes);
+
+// Function to obtain heuristic results for viable matmul algorithms
+// Note: in dnncompat this function will always return a single heuristic
+Status_t MatmulAlgoGetHeuristic(Handle_t handle,
+                                MatmulDesc_t operationDesc,
+                                MatrixLayout_t Adesc,
+                                MatrixLayout_t Bdesc,
+                                MatrixLayout_t Cdesc,
+                                MatrixLayout_t Ddesc,
+                                MatmulPreference_t preference,
+                                int requestedAlgoCount,
+                                MatmulHeuristicResult_t heuristicResultsArray[],
+                                int* returnAlgoCount);
+```
+
+###### Notes
+
+- Several features such as `MatmulPreference_t`, `MatmulPreferenceCreate`,
+  `MatmulPreferenceSetAttribute`, and `ReductionScheme_t` are not mappable to
+  oneDNN as they are internal to the implementation and not exposed to users
+- To work around the absence of direct `Algo_t` mapping, the heuristic result
+  structure can be adapted. Instead of directly mapping `Algo_t`, a new struct
+  containing the oneDNN primitive descriptor could be used. This allows the
+  primitive to be initialized early, the required workspace size to be queried,
+  and gives the user control over the memory allocation for the workspace.
+
+## Caveats
+
+While `dnncompat` aims to provide an API as close as possible to vendor-specific
+libraries, there are some caveats:
+
+- There are differences in the range of operators supported by oneDNN compared
+  to vendor-specific libraries.
+
+- There are also differences in feature support. Some operation features present
+  in vendor-specific libraries are not available or may function differently in
+  oneDNN and vice versa.
+
+- For operations such as Activation and Pooling, cuDNN allows the option to
+  disable `NaN` propagation. However, in oneDNN, `NaN` values are always
+  propagated by default, which could lead to different outcomes in the execution
+  of similar operations. In the initial implementation a warning will be printed
+  when `NaN` propagation is explicitly disabled.
+
+- Multiple Handle Support: cuDNN supports the creation and usage of multiple
+  handles. In `dnncompat`, while multiple handles can be created, they must all
+  be initialized with the same `sycl::queue` as the primary handle. This
+  limitation will lead to a slight difference in usage of `dnncompat` in some
+  applications.
+
+- cuDNN includes a specialized FFT convolution implementation, while this
+  algorithm is not supported in oneDNN.
```

Comment:
what should dnncompat behavior be here? print a warning and move forward with direct convolution algorithm?